cli_tools/common/utils/logging/service/cli_tool_with_log.go

```diff
@@ -0,0 +1,46 @@
+//  Copyright 2019 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package service
+
+import (
+	"flag"
+	"os"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+)
+
+// CliToolWithLogging abstracts the interface of a cli tool with logging. A tool
+// which implemented this interface adopts log service automatically.
+type CliToolWithLogging interface {
+	// ActionType indicates the action type of the tool. It represents which tool
+	// function this log line is about. "ImageImport" is an example.
+	ActionType() ActionType
+
+	// InitParamLog initializes tool's input params in order to be logged
+	InitParamLog() InputParams
+
+	// Run is the entry function of the tool
+	Run() (*daisy.Workflow, map[string]string, error)
```

Comment:
The usage of map[string]string here and elsewhere feels dangerous for two reasons:
1. There's no contract that the caller or the receiver can agree upon.
2. The map is mutable, so it's unclear:
   a. Who is responsible for setting the fields
   b. Who is allowed to mutate the fields

Reply:
Agreed, all that I tried to avoid is to prevent reusing strcuts from logging.service package. I have already used a struct from param package to resolve this problem in my latest commit.

Reply:
Need to work more on "mutable" thing. That may imply the ideal solution.